Separate http1 rx and tx state machines to support pipelining.

jrcheli opened the pull request:
resolves #787 , in part.

michalbiesek commented on src/state_private.h:
```diff
@@ -176,7 +176,7 @@ typedef struct net_info_t {
     int active;
     int type;
-    http_state_t http;
+    http_state_t http[2];      // rx=[0] and tx=[1]
```
nit: I know this is not related to your changes, but can we use an enum here instead of magic values like `0`, `1`, `2`.
IMHO it will improve readability, like in the code places below:
```
        resetHttp(&ninfo->http[0]);
        resetHttp(&ninfo->http[1]);
```
When I read the code above I need to check if `0` described RX or TX here in `state.private.h`

michalbiesek commented on src/httpstate.c:
```diff
@@ -45,17 +45,14 @@ setHttpState(http_state_t *httpstate, http_enum_t toState)
+            httpstate->http2Buf.len = 0;
+            httpstate->http2Buf.size = 0;
+
+            if (httpstate->hdr) free(httpstate->hdr);
```
nit: SInce you added the freeing and NULL part above can you use same schema here, e.g:
instead of using:
```
if (httpstate->hdr) free(httpstate->hdr);
httpstate->hdr = NULL;
```
call
```
if (httpstate->hdr) {
  free(httpstate->hdr);
  httpstate->hdr = NULL;
}
```
In other words, set `NULL` value to `httpstate->hdr` only in case `httpstate->hdr` was something different than `NULL` before.

jrcheli commented:
@michalbiesek, I reverted my changes, and made them again from scratch.
I *think* I addressed all of your comments that didn't become obsolete during this process.  If I haven't don't hesitate to poke me again.

michalbiesek commented on src/httpstate.c:
```diff
@@ -673,7 +670,7 @@ parseHttp2(http_state_t* state, net_info *net, int isTx,
+    http_buf_t    *stash  = &state->http2Buf;       // stash for partial frames
     while (bufLen > 0) {
         // skip over MAGIC
         if (bufLen >= HTTP2_MAGIC_LEN && !strncmp((char*)buf, HTTP2_MAGIC, HTTP2_MAGIC_LEN)) {
```
@jrcheli just FYI in the code mentioned above in function `parseHttp2` the `isTx` argument is utilized only for the logging part which is now commented:
```
        //scopeLogDebug("DEBUG: HTTP/2 %s frame found; type=0x%02x, flags=0x%02x, stream=%d",
        //        isTx ? "TX" : "RX", fType, fFlags, fStream);
```
In other words `isTx` is an unused argument now. You can leave it as it is - I just want to let you know about it.
Discussion resolved.

michalbiesek commented:
LGTM, regarding the [comment](https://github.com/criblio/appscope/issues/787#issuecomment-1048140617).

I use my demo setup to test it:
### Scenario 1
```
apt update
```
-  before the changes from PR 809 (version 1.0.0):
GET (a valid ones): 85% of total http_methods
garbage value (invalid ones): 15% of total http_methods

-  after the changes from PR 809 (sha 391fdc):
GET (a valid ones): 100% of total http_methods
garbage value (invalid ones): 0 of total http_methods

### Scenario 2 (done after Scenario 1 but excluding data received from Scenario 1)
```
apt install curl wget gdb gcc clang nano htop wireshark -y
```
-  before the changes from PR 809 (version 1.0.0):
GET (a valid ones): 27% of total http_methods
garbage value (invalid ones): 73% of total http_methods

-  after the changes from PR 809 (sha 391fdc):
GET (a valid ones): 85% of total http_methods
garbage value (invalid ones): 15% of total http_methods

jrcheli commented:
Merging.  As @michalbiesek confirmed above, this isn't resolving *everything*, but is moving things forward.  #829 has been opened to help account for the fact that this PR not a complete fix for the original observation.
